src/Driver/PDO/Connection.php

```diff
@@ -87,17 +91,80 @@ public function quote(string $value): string
         return $this->connection->quote($value);
     }
 
-    public function lastInsertId(?string $name = null): string
+    /**
+     * {@inheritDoc}
+     */
+    public function lastInsertId()
     {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'oci') {
+            throw IdentityColumnsNotSupported::new();
+        }
+
         try {
-            if ($name === null) {
-                return $this->connection->lastInsertId();
+            $lastInsertId = $this->connection->lastInsertId();
+        } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws a 'lastval is not yet defined in this session' error when no identity value is
+            // available, with SQLSTATE 55000 'Object Not In Prerequisite State'
+            if ($sqlState === '55000' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw NoIdentityValue::new($exception);
             }
 
-            return $this->connection->lastInsertId($name);
+            throw Exception::new($exception);
+        }
+
+        if (! $this->isValidLastInsertIdOrSequenceNumber($lastInsertId)) {
+            throw NoIdentityValue::new();
+        }
+
+        return $lastInsertId;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSequenceNumber(string $name)
+    {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'mysql') {
+            throw SequencesNotSupported::new();
+        }
+
+        if ($driverName === 'oci') {
+            throw new SequencesNotSupported('Oracle supports sequences, but the PDO OCI driver does not support them.');
+        }
+
+        try {
+            $sequenceNumber = $this->connection->lastInsertId($name);
         } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws an 'Undefined table' when the sequence does not exist
+            if ($sqlState === '42P01' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw SequenceDoesNotExist::new($name, $exception);
+            }
+
             throw Exception::new($exception);
         }
+
+        if (! $this->isValidLastInsertIdOrSequenceNumber($sequenceNumber)) {
+            throw SequenceDoesNotExist::new($name);
+        }
+
+        return $sequenceNumber;
+    }
+
+    /**
+     * @param int|string $value
+     */
+    private function isValidLastInsertIdOrSequenceNumber($value): bool
+    {
+        // pdo_mysql & pdo_sqlite return 0 or '0', pdo_sqlsrv returns ''
```

Comment:
Looks like a lot of driver-specific logic in a driver-agnostic place. I'm curious if it can have false-positive effects. E.g. an auto-incremented field is incremented form `-1` to `0`.

Just a thought, would it be safer if we only failed if it's not a numeric string? Otherwise, if neither the user nor the DB driver care about the proper logic, why should we?

Reply:
> E.g. an auto-incremented field is incremented form -1 to 0.

I'm not sure if any DB allows negative auto-increments? MySQL at leasts fails:

```
ALTER TABLE ai AUTO_INCREMENT=-1;
#1064 - You have an error in your SQL syntax (...)
```

> Otherwise, if neither the user nor the DB driver care about the proper logic, why should we?

Consistency. If one driver fails with an exception when a value is not available, I think all of them should. If this driver-specific logic is a problem, maybe the individual checks can be moved to the driver-specific connection class: `PDO\SQLSrv\Connection` would only check for an empty string, while others would only check for `0`, etc.?

Reply:
> Consistency.

This may provide a false sense of consistency. As it's already pointed out in the documentation, it is recommended to use this method right after the insert, otherwise the behavior depends on the platform. Why do we want to enforce consistency in the case which is not recommended and where full consistency is not possible?

Reply:
Getting an outdated value (i.e. for an older INSERT) is one thing we can't avoid, and there will be inconsistencies between drivers.

But getting an empty- or zero-value is something that, IMO, should consistently throw an exception. You *do* want your code to fail if for any reason, you're calling this method after an INSERT to a table that doesn't have an auto-increment field: you don't want it to silently continue with a zero value.

Reply:
I agree. My point was that a zero value is only possible in the case of a logical error, which we declare to not handle consistently. Even this zero value may or may not happen depending on the currently used driver, so we won't be able to throw this exception consistently with the same invalid logic but different drivers.

I'm not strongly opposed to handling zero values, I'm just trying to come up with a clear understanding of what our API contract is.

Reply:
I guess our API contract is:

- you are not guaranteed to get a value if you don't call `getLastInsertId()` right after an `INSERT` statement
- if you do not get a value, you get an exception

Which the current code fulfills well.